docs/ref/models/querysets.txt

```diff
@@ -264,6 +264,60 @@ control the name of the annotation::
 For an in-depth discussion of aggregation, see :doc:`the topic guide on
 Aggregation </topics/db/aggregation>`.
 
+
+.. currentmodule:: django.db.models.query
+
+.. class:: FilteredRelation(relation_name, condition=Q())
+
+    .. versionadded:: 2.0
+
+Annotate supports also a new kind of argument that is an instance of
+:class:`.FilteredRelation` .
+This kind of parameter allows to pass filtering options inside the the ``ON``
+clause when a ``JOIN`` is performed.
+It does not act on the default relationship but on the given ``alias``.
+
+Example::
+
+    >>> from django.db.models import FilteredRelation
+    >>> Restaurant.objects.annotate(
+    ...    pizzas_vegetarian=FilteredRelation(
+    ...        'pizzas', condition=Q(pizzas__vegetarian=True),
+    ...    )).filter(pizzas_vegetarian__name__icontains='Mozzarella')
+
+This query will join the model Restaurant only to Pizzas that have vegetarian
+flag to True. This results into a performance improvement, noticeable if the
+join is performed through a large number of rows. Because the filtering from
+the ``WHERE`` clause will only operate on a subset of Pizzas (the vegetarian
+ones).
+
+.. admonition:: You need to be aware of some limitations
+
+    The :class:`.FilteredRelation` can not contain
+    conditions that span over deep relations.
+    The condition can only apply on fields that satisfy the constraint
+    ``Field.is_relation is False``.
+
+    Example::
+
+        >>> Restaurant.objects.annotate(
+        ...    pizzas_with_toppings_startswith_n=FilteredRelation(
+        ...        'pizzas__toppings',
+        ...        condition=Q(pizzas__toppings__name__startswith='n'),
+        ...     )).filter(pizzas_with_toppings_startswith_n__isnull=False)
+
+    Will raise ``ValueError`` "Filtered relation
+    'pizzas_with_toppings_startswith_n' cannot operate on foreign key
+    'pizzas__toppings'".
+
+.. admonition:: :meth:`~.QuerySet.prefetch_related()` is not yet supported
+
+    We might consider adding support of
+    :meth:`~.QuerySet.prefetch_related()` in combination of
+    :class:`.FilteredRelation` in the future.
```

Comment:
I don't think the docs should promise things, because there's a chance it won't be implemented and this message will be left there for years. Just explain that they don't work together.